trieste/acquisition/function.py

```diff
@@ -130,6 +133,120 @@ def expected_improvement(model: ModelInterface, eta: tf.Tensor, at: QueryPoints)
+            and approximately generate the samples of the minimum (recommend scaling with problem dimension).
+        """
+        self._num_samples = num_samples
+        self._grid_size = grid_size
+        self._search_space = search_space
+
+    def prepare_acquisition_function(
+        self, dataset: Dataset, model: ModelInterface
+    ) -> AcquisitionFunction:
+        """
+        Need to sample possible min-values from our posterior. To do this we implement a Gumbel sampler.
+        we approximate Pr(y*^hat<y) by Gumbel(alpha,beta) then sample from Gumbel.
+
+        :param dataset: Unused.
+        :param model: The model over the specified ``dataset``.
+        :return: The MES function.
+        """
+
+        # generate random grid to fit Gumbel distribution
+        # grid consists of already queried locations and grid_size other random locations
+        query_points = self._search_space.sample(self._grid_size)
+        query_points = tf.concat([dataset.query_points,query_points],0)
+        fmean, fvar = model.predict(query_points)
+        fsd = tf.math.sqrt(fvar)
+        idx = tf.math.argmax(-fmean[:tf.shape(dataset.query_points)[0]])
+
+        # Fit Gumbel distriubtion
+        # scaling so that gumbel scale is proportional to IQ range of cdf Pr(y*<z)
+        # find quantiles Pr(y*<y1)=r1 and Pr(y*<y2)=r2
+
+        right = tf.squeeze(fmean[tf.squeeze(idx)])
+        left = right
+        probf = lambda x: tf.math.exp(tf.math.reduce_sum(tfp.distributions.Normal(tf.cast(0,tf.float64),tf.cast(1,tf.float64)).log_cdf(-(x - fmean) / fsd), axis=0))
+        
+
+        # get range for binary search
+        i = 0
+        while probf(left) < 0.75:
+            left = 2. ** i * tf.math.reduce_min(fmean - 5. * fsd) + (1. - 2. ** i) * right
+            i += 1
+        i = 0
+        while probf(right) > 0.25:
+            right = -2. ** i * tf.math.reduce_min(fmean - 5. * fsd) + (1. + 2. ** i) * tf.squeeze(fmean[tf.squeeze(idx)])
+            i += 1
+
+
+        # Binary search for 3 percentiles
+        q1, med, q2 = map(lambda val: bisect(lambda x: probf(x) - val, left, right, maxiter=10000, xtol=0.00001),
+                            [0.25, 0.5, 0.75])
+
+
+        # solve for gumbel params
+        beta = (q1 - q2) / (tf.math.log(tf.math.log(4. / 3.)) - tf.math.log(tf.math.log(4.)))
+        alpha = med + beta * tf.math.log(tf.math.log(2.))
+
+        # sample K length vector from unif([0,1])
+        # return K Y* samples
+        samples = -tf.math.log(-tf.math.log(tf.random.uniform([self._num_samples],dtype=tf.dtypes.float64))) * tf.cast(beta,tf.float64) + tf.cast(alpha,tf.float64)
+
+
+        return lambda at: self._acquisition_function(model, samples, at)
+
+    @staticmethod
+    def _acquisition_function(model: ModelInterface, samples: tf.Tensor, at: QueryPoints) -> tf.Tensor:
+        return mes(model, samples, at)
+
+
+def mes(model: ModelInterface, samples: tf.Tensor, at: QueryPoints) -> tf.Tensor:
```

Comment:
replace with an explicit name, eg maxvalue_entropy_change (or ..._reduction, or any better name)?

Reply:
agreed. I think `MES -> MaxValueEntropySearch` and `mes` -> `max_value_entropy_search` would be good. What do you reckon?